elkscmd/man/man1/dd.1
.TH DD 1
.SH NAME
dd \- convert and copy a file
.SH SYNOPSIS
\fBdd\fR [\fIoption = value\fR] ...\fR
.br
.SH EXAMPLES
.TP 30
.B dd if=/dev/fd0 of=/dev/fd1
# Copy disk 0 to disk 1
.TP 30
.B dd if=x of=y bs=1w skip=4
# Copy \fIx\fP to \fIy\fP, skipping 4 words
.TP 30
.B dd if=x of=y count=3
# Copy three 512\-byte blocks
.SH DESCRIPTION
.PP
This command is intended for copying partial files.
The block size, skip count, and number of blocks to copy can be specified.
The options are:
.PP
.ta 0.25i 1.5i
	\fBif\fR = file	\- Input file (default is \fIstdin\fR)
.br
	\fBof\fR = file	\- Output file (default is standard output)
.br
	\fBibs\fR = n	\- Input block size (default 512 bytes)
.br
	\fBobs\fR = n	\- Output block size (default is 512 bytes)
.br
	\fBbs\fR = n	\- Block size; sets \fIibs\fP and \fIobs\fP (default is 512 bytes)
.br
	\fBskip\fR = n	\- Skip \fIn\fP input blocks before reading
.br
	\fBseek\fR = n	\- Skip \fIn\fP output blocks before writing
.br
	\fBcount\fR = n	\- Copy only \fIn\fP input blocks
.br
	\fBconv = lcase\fR	\- Convert upper case letters to lower case
.br
	\fBconv = ucase\fR	\- Convert lower case letters to upper case
.br
	\fBconv = swab\fR	\- Swap every pair of bytes
.br
	\fBconv = noerror\fR	\- Ignore errors and just keep going
.br
	\fBconv = notrunc\fR	\- Do not truncate unmodified blocks
.br
	\fBconv = silent\fR	\- Suppress statistics (MINIX 3 specific flag)
.PP
Where sizes are expected, they are in bytes.
However, the letters \fBw\fR, \fBb\fR, or \fBk\fR may be appended to the
number to indicate words (2 bytes), blocks (512 bytes), or K
(1024 bytes), respectively.
When
.I dd
is finished, it reports the number of full and partial blocks read and written.


elks/CHANGELOG
This file is used to record changes made to the CVS database. Please add to
the top of this file before committing any changes. Please use the header
format provided for each entry, noting that the text on each line should not
be longer than the lines of dashes.

Al Riddoch.

Contributor           Email address                     CHANGELOG entries found
~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~
Alan Cox              alan@redhat.com                   17 Jun 2001-18 Jun 2001
Alastair Bridgewater  nyef@softhome.net                 19 Nov 1998- 6 Dec 1998
Alistair Riddoch      alriddoch@zepler.org              25 Sep 1998-17 Jun 2001
Arnaud Launay         asl@launay.org                    15 Jul 1999-17 Oct 2001
Christian Mardmuller  chm@kdt.de                        10 Nov 1999-29 Feb 2000
Christoph Plattner    Christoph.Plattner@dot.at         31 Jan 2000- 2 Feb 2000
Claudio Matsuoka      claudio@conectiva.com
Daniele Giacomini     daniele@swlibero.org		20 Sep 2004
David Murn            scuffer@hups.apana.org.au         11 Nov 1998-15 Nov 1999
Eduardo Habkost	      ehabkost@conectiva.com.br		01 Jun 2004
Giles Russell                                           22 Nov 1999
Greg Haerr                                              14 Jul 1999-20 Jul 1999
Harry Kalogirou       HarKal@gmx.net                    20 Jun 2001- 2 Jan 2003
Jacek Lipkowski       sq5bpf@acid.ch.pw.edu.pl          20 Jan 2002
Jay Barr                                                27 Oct 2002
Ken Martwick          kenm@efn.org                      12 May 2002
M Warner Losh         imp@village.org                   20 Jan 2002
Manuel Novoa III      mjn3@codepoet.org                 18 Jun 2002-25 Jun 2002
Mario Frasca                                            20 Jan 2000-14 Feb 2000
Miguel Bolanos	      mike@linuxlabs.com		20 Sep 2004
Paul Nasrat           pauln@truemesh.com                29 Oct 2001-12 Jul 2002
Per Olofsson          MagerValp@cling.gu.se             25 Feb 2000
Peter L Vachuska      pvachusk@nconnect.net             13 Aug 2001
Richard Wallman       murkygoth@murkygoth.uklinux.net    2 Apr 2002
Riley Williams        Riley@Williams.Name               16 Nov 1998-13 Jul 2003
Roman Festchook       roma@polesye.net                  27 Oct 2001
Simon Weijgers        simon@weijgers.com                16 Sep 1998-15 Nov 1999
Simon Wood                                               6 Jan 2000-12 Jan 2000
Thomas McWilliams     tgm9999@yahoo.com                 27 Aug 1999- 2 Apr 2001
Tommy McCabe 	      rocketjet314@yahoo.com		31 May 2004

-------------------------------------------------------------------------------

Thu Feb 16 17:03:00 EST 2012	Jody Bruchon <jody@jodybruchon.com>	
	* Removed elksfs. It was a copy of the MINIX fs and no tools
	  exist to make such a filesystem.

Thu Nov 23 21:35:28 EST 2006	Segin <segin2005@gmail.com>
	* Changed name for the submenu for selecting charater
	  device drivers to note the fact that that submenu is broken;
	  attempting to enter it will have lxdialog die horribly.

Wed Nov  8 17:49:52 EST 2006	Segin <segin2005@gmail.com>

	* Changed ELKS init sequence to check all locations for 'init'
	  that the Linux kernel does. This is to follow more 
	  "traditional" behaviour.

Sun Aug 20 21:51:25 UTC 2006	Claudio Matsuoka <claudio@mandriva.com>

	* Allow loading Minix executables with total size larger than
	  64KB, such as ash (70140 bytes). Currently the total size
	  minus 32 bytes of header was limited to 16-bit.

	* Set tty1 back to minor 0. This is different from Linux
	  numbering, but it's necessary to have current ELKS virtual
	  terminals working. This and the executable size change put
	  ELKS back in usable state, allowing logins in the first VT
	  and executing a shell.

Mon Sep 20 23:27:23 CST 2004	Daniele Giacomini <daniele@swlibero.org>

	* Added italian keys support, Miguel Bolanos cleaned the original
	  file code.

Wed Jun  9 17:27:52 BRT 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* fix for my last fix: s/rp_cmd/rq_cmd/

Thu Jun 03 21:47:13 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* fixed format specifiers to correctly displaysegment sizes as unsigned
	  values, and to display debug messages correctly.

Thu Jun 03 07:48:25 CST 2004	Eduardo Habkost <ehabkost@conectiva.com.br>

	* Fixed wrong struct field name for arch/i86/drivers/block/doshd.c
	* Several fixes ro arch/i86/mm/user.c

Web Jun 02 14:16:57 CST 2004    Claudio Matsuoka <claudio@conectiva.com>

	* Added an extra fix to Eduardos memcpy fix.

Tue Jun 01 21:00:03 CST 2004	Eduardo Habkost ehabkost@conectiva.com.br

	* Fix memcpy_{to,from}fs, adding code to save the contents of SI and DI

Mon May 31 07:39:47 CST 2004	Tommy McCabe <rocketjet314@yahoo.com>

	* Fix wrong assignment (typo?) on printk
	* Fix order of LD parameters on arch/i86/Makefile
	* Use #ifdef rightly on setup.S
	* Fix init to use tty0, not tty1 as system console.

===============================================================================

Sun Jul 13 04:24:44 BST 2003	Riley Williams <Riley@Williams.Name>

	* This is ELKS 0.1.3-pre1

Sat Jul 12 12:24:36 BST 2003	Riley Williams <Riley@Williams.Name>

	* Rewrote kernel/printk.c to allow printk() to correctly print out
	  signed values.
	* Rewrote kernel/printk.c to produce a less verbose panic() dump,
	  and also to increase the number of lines of stack trace to 10.
	* Various spelling corrections to printk() messages.

===============================================================================

Thu Jan 02 16:12:36 EET 2003	Harry Kalogirou <harkal@gmx.net>

        * Better select hash.
	* This is ELKS 0.1.2

-------------------------------------------------------------------------------

Fri Nov 01 01:52:44 EET 2002	Harry Kalogirou <harkal@gmx.net>

        * Changed the position of ds in the _registers struct.
        * Added support for loading the new ELKS executable format. This is 

elkscmd/file_utils/dd.c
/*
 * Copyright (c) 1993 by David I. Bell
 * Permission is granted to use, distribute, or modify this source,
 * provided that this copyright notice remains intact.
 *
 * The "dd" command - adapted from busybox/busyelks.
 */

#include "futils.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define	PAR_NONE	0
#define	PAR_IF		1
#define	PAR_OF		2
#define	PAR_BS		3
#define	PAR_COUNT	4
#define	PAR_SEEK	5
#define	PAR_SKIP	6


struct param {
	char	*name;
	int	value;
};

static struct param params[] = {
	{ "if",		PAR_IF },
	{ "of",		PAR_OF },
	{ "bs",		PAR_BS },
	{ "count",	PAR_COUNT },
	{ "seek",	PAR_SEEK },
	{ "skip",	PAR_SKIP },
	{ NULL,		PAR_NONE }
};


/* Fixed buffer */
static char localbuf[BUFSIZ];		/* use disk block size for efficiency*/


/*
 * Read a number with a possible multiplier.
 * Returns -1 if the number format is illegal.
 */
static long getnum(const char *cp)
{
	long value;

	if (!isdecimal(*cp))
		return -1;

	value = 0;
	while (isdecimal(*cp))
		value = value * 10 + *cp++ - '0';

	switch (*cp++) {
		case 'k':
			value *= 1024;
			break;

		case 'b':
			value *= 512;
			break;

		case 'w':
			value *= 2;
			break;

		case '\0':
			return value;

		default:
			return -1;
	}

	if (*cp) return -1;

	return value;
}


int main(int argc, char **argv)
{
	char	*str;
	char	*cp;
	struct param	*par;
	char	*infile;
	char	*outfile;
	int	infd;
	int	outfd;
	int	incc;
	int	outcc;
	int	blocksize;
	long	count = -1;
	long	seekval;
	long	skipval;
	long	intotal = 0;
	long	outtotal = 0;
	char	*buf;
	int	retval = 1;

	infile = NULL;
	outfile = NULL;
	blocksize = 512;
	skipval = 0;
	seekval = 0;

	while (--argc > 0) {
		str = *++argv;
		cp = strchr(str, '=');
		if (cp == NULL) {
			fprintf(stderr, "Missing or invalid argument(s)\n");
			goto usage;
		}
		*cp++ = '\0';

		for (par = params; par->name; par++) {
			if (strcmp(str, par->name) == 0)
				break;
		}

		switch (par->value) {
			case PAR_IF:
				if (infile) {
					fprintf(stderr, "Multiple input files illegal\n");
					goto usage;
				}
	
				infile = cp;
				break;

			case PAR_OF:
				if (outfile) {
					fprintf(stderr, "Multiple output files illegal\n");
					goto usage;
				}

				outfile = cp;
				break;

			case PAR_BS:
				blocksize = getnum(cp);
				if (blocksize <= 0) {
					fprintf(stderr, "Bad block size value\n");
					goto usage;
				}
				break;

			case PAR_COUNT:
				count = getnum(cp);
				if (count < 0) {
					fprintf(stderr, "Bad count value\n");
					goto usage;
				}
				break;

			case PAR_SEEK:
				seekval = getnum(cp);
				if (seekval < 0) {
					fprintf(stderr, "Bad seek value\n");
					goto usage;
				}
				break;

			case PAR_SKIP:
				skipval = getnum(cp);
				if (skipval < 0) {
					fprintf(stderr, "Bad skip value\n");
					goto usage;
				}
				break;

			default:
				fprintf(stderr, "Unknown dd parameter\n");
				goto usage;
		}
	}

	if (infile == NULL) {
		infile = "-";
	}

	if (outfile == NULL) {
		outfile = "-";
	}

	buf = localbuf;
	if (blocksize > sizeof(localbuf)) {
		buf = malloc(blocksize);
		if (buf == NULL) {
			fprintf(stderr, "Cannot allocate buffer\n");
			return retval;
		}
	}

	if (!strcmp(infile, "-")) 
		infd = 0;
	else {
		infd = open(infile, 0);
		if (infd < 0) {
			perror(infile);
			if (buf != localbuf) free(buf);
			goto usage;
		}
	}

	if (!strcmp(outfile, "-"))
		outfd = 1;
	else {
		outfd = creat(outfile, 0666);
		if (outfd < 0) {
			perror(outfile);
			close(infd);
			if (buf != localbuf) free(buf);
			goto cleanup2;
		}
	}

	if (skipval) {
		if (lseek(infd, skipval * blocksize, 0) < 0) {
			while (skipval-- > 0) {
				incc = read(infd, buf, blocksize);
				if (incc < 0) {
					perror(infile);
					goto cleanup;
				}

				if (incc == 0) {
					fprintf(stderr, "Skipped beyond end of file\n");
					goto cleanup;
				}
			}
		}
	}

	if (seekval) {
		if (lseek(outfd, seekval * blocksize, 0) < 0) {
			perror(outfile);
			goto cleanup;
		}
	}

	/* If count is specified, only copy that many blocks */
	if (count > 0) 
		count *= blocksize;
	else 
		if (count < 0) count = 0x7fffffff;
	else 
		goto cleanup;	/* exit immediately if count == 0 */

	while ((count > intotal) && (incc = read(infd, buf, blocksize)) > 0) {
		intotal += incc;
		cp = buf;

		while (incc > 0) {
			outcc = write(outfd, cp, incc);
			if (outcc < 0) {
				perror(outfile);
				goto cleanup;
			}

			outtotal += outcc;
			cp += outcc;
			incc -= outcc;
		}
	}

	/* Exit status can only become 0 (no error) at this point */
	if (incc < 0) 
		perror(infile);
	else 
		retval = 0;

cleanup:
	if (close(outfd) < 0) perror(outfile);
cleanup2:
	close(infd);
	if (buf != localbuf) free(buf);
	fprintf(stderr, "%ld+%d records in\n", intotal / blocksize,
		(intotal % blocksize) != 0);
	fprintf(stderr, "%ld+%d records out\n", outtotal / blocksize,
		(outtotal % blocksize) != 0);
	fprintf(stderr, "%ld bytes (%ld%c KiB) copied\n", outtotal, outtotal >> 10, (outtotal & 0x3ff) ? '+' : '\0');
	return retval;

usage:
	fprintf(stderr, "\nUsage: dd if=<inflie> of=<outfile> [optional_params ...]\n\n");
	fprintf(stderr, "If if= and/or of= are omitted, stdin/stdout will be used.\n");
	fprintf(stderr, "Optional parameters:\n");
	fprintf(stderr, "bs=<blocksize>  seek=<count>  skip=<count>  count=<count>\n");
	fprintf(stderr, "seek/skip skips <count> blocks in input/output files, respectively\n");
	fprintf(stderr, "count copies only <count> blocks (default is until end of file)\n\n");
	return retval;
}
